### PR TITLE
Reusable storage bridge

### PR DESCRIPTION
**Pull request by jason-bragg**
Refactored storage bridge and stateful grain handling to allow more flexable state management needed by transaction logic.
- Removed grain state managment from catalog and grain creator.
- Grain<> and LogConsistancy grain now setup their own state via the lifecycle.
- GrainStateStorageBridge has been replaced with StateStorageBridge which can be used for transaction and storage facet implementations which choose to store state using storage providers.

Note: With this change, only type managment, and statistics issues prevent event sourcing and stateful grain (grain<>) from being moved out of orleans core/runtime into orleans extension libraries.

See Orleans Transactions #3369

### CODE REVIEW

File: src/OrleansRuntime/Catalog/GrainCreator.cs [hunk 1]
```diff
@@ -47,56 +38,13 @@ public Grain CreateGrainInstance(IGrainActivationContext context)
...
-
-            if (statefulGrain == null)
-                return grain;
-
-            //Inject state and storage data into the grain
-            statefulGrain.GrainState.State = Activator.CreateInstance(stateType);
-            statefulGrain.SetStorage(storage);
+            // wire up to lifecycle
+            var participant = grain as ILifecycleParticipant<GrainLifecycleStage>;
+            participant?.Participate(context.ObservableLifecycle);
```
Comment (jason-bragg) [inline src/OrleansRuntime/Catalog/GrainCreator.cs]: Moving state management logic out of catalog and grain-creator into grain base classes that participate in the lifecycle is the bulk of this change.  While we should be moving away from base classes, this preserves the existing behaviors and moves this logic out of the catalog/grain-creator and provides a more extensible pattern for future features.

File: src/Orleans/Core/Grain.cs [hunk 2]
```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
         {
             return storage.ReadStateAsync();
         }
+
+        public void Participate(ILifecycleObservable<GrainLifecycleStage> lifecycle)
+        {
+            lifecycle.Subscribe(GrainLifecycleStage.SetupState, this);
+        }
+
+        public async Task OnStart(CancellationToken ct)
```
Comment (ReubenBond) [inline src/Orleans/Core/Grain.cs]: suggest we make this an explicit implementation so as not to pollute intellisense for Grain

File: src/OrleansRuntime/Hosting/DefaultSiloServices.cs [hunk 3]
```diff
@@ -61,8 +63,15 @@ internal static void AddDefaultServices(IServiceCollection services)
             services.TryAddTransient(typeof(IStreamSubscriptionObserver<>), typeof(StreamSubscriptionObserverProxy<>));
 
             services.TryAddSingleton<StatisticsProviderManager>();
+
             services.TryAddSingleton<StorageProviderManager>();
+            services.TryAddSingleton<IKeyedServiceCollection<string, IStorageProvider>>(sp => sp.GetRequiredService<StorageProviderManager>());
```
Comment (ReubenBond) [inline src/OrleansRuntime/Hosting/DefaultSiloServices.cs]: Doesn't need to change in this PR, but there's an extension method called `TryAddFromExisting` which essentially does this.

**Comment by ReubenBond** [comment]
One change to consider is that when a lifecycle participant fails, propagate the original exception to the caller instead of an `OperationCanceledException`.

Highlighted in the failure of `Persistence_Grain_Activate_Error`:
`MESSAGE:
Assert.Throws() Failure\r\nExpected: typeof(UnitTests.StorageTests.StorageProviderInjectedError)\r\nActual: typeof(System.OperationCanceledException): Lifecycle start canceled due to errors at stage SetupState`

That change had no effect until this PR, but I believe it was already in place.

File: src/Orleans/Core/Grain.cs [hunk 4]
```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
...
+            lifecycle.Subscribe(GrainLifecycleStage.SetupState, this);
+        }
+
+        async Task ILifecycleObserver.OnStart(CancellationToken ct)
+        {
+            if (ct.IsCancellationRequested)
+                return;
+            IStorageProvider storageProvider = GetStorageProvider();
+            string grainTypeName = this.GetType().FullName;
+            this.storage = new StateStorageBridge<TGrainState>(this.GetType().FullName, this.GrainReference, storageProvider);
```
Comment (jdom) [inline src/Orleans/Core/Grain.cs]: you probably meant to use `grainTypeName` as a parameter here

File: src/Orleans/Core/Grain.cs [hunk 5]
```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
         {
             return storage.ReadStateAsync();
         }
+
+        void ILifecycleParticipant<GrainLifecycleStage>.Participate(ILifecycleObservable<GrainLifecycleStage> lifecycle)
```
Comment (jdom) [inline src/Orleans/Core/Grain.cs]: Implementing this explicitly would mean that app devs (or infrastructure extensions) will not be able to create a base class that is a participant and hook it up as expected, correct? Maybe it should just be protected virtual instead?

File: src/Orleans/Core/Grain.cs [hunk 6]
```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
...
+        {
+            lifecycle.Subscribe(GrainLifecycleStage.SetupState, this);
+        }
+
+        async Task ILifecycleObserver.OnStart(CancellationToken ct)
```
Comment (jdom) [inline src/Orleans/Core/Grain.cs]: maybe the `ILifecycleObserver` itself can be instantiated on the `Participate` phase, so that the grain itself doesn't have to implement `ILifecycleObserver` that could conflict with how the users implement their grains?
For example the `StateStorageBridge` itself could implement that interface, or an inner class of it, or even have it implicitly implemented with a lambda that can use closures.

File: src/Orleans/Core/Grain.cs [hunk 7]
```diff
@@ -332,5 +324,51 @@ protected virtual Task ReadStateAsync()
...
+        {
+            if (ct.IsCancellationRequested)
+                return;
+            IStorageProvider storageProvider = GetStorageProvider();
+            string grainTypeName = this.GetType().FullName;
```
Comment (jdom) [inline src/Orleans/Core/Grain.cs]: I think this is not backwards compatible. In the past we were using `$"{grainType}[{genericArguments}]` as the state name, where grain type is not even `type.FullName` or something like that.

File: src/OrleansRuntime/Catalog/Catalog.cs [hunk 8]
```diff
@@ -1349,6 +1142,11 @@ private void RerouteAllQueuedMessages(ActivationData activation, ActivationAddre
             {
                 RequestContext.Import(requestContextData);
                 await activation.Lifecycle.OnStart();
+                // TODO: Consider pre and post activation stages in lifecycle for PreActivateProtocolParticipant and PostActivateProtocolParticipant like behaviors
+                if (activation.GrainInstance is ILogConsistencyProtocolParticipant)
+                {
+                    await ((ILogConsistencyProtocolParticipant)activation.GrainInstance).PreActivateProtocolParticipant();
```
Comment (jdom) [inline src/OrleansRuntime/Catalog/Catalog.cs]: question: shouldn't this happen before calling `activation.Lifecycle.OnStart();`

File: src/OrleansRuntime/Catalog/Catalog.cs [hunk 9]
```diff
@@ -1349,6 +1142,11 @@ private void RerouteAllQueuedMessages(ActivationData activation, ActivationAddre
             {
                 RequestContext.Import(requestContextData);
                 await activation.Lifecycle.OnStart();
+                // TODO: Consider pre and post activation stages in lifecycle for PreActivateProtocolParticipant and PostActivateProtocolParticipant like behaviors
```
Comment (jdom) [inline src/OrleansRuntime/Catalog/Catalog.cs]: Yes, I was even considering that the `Grain.OnActivateAsync` could actually be implicitly hooked up by the `Grain` base class being a lifecycle participant, not as a special method that we handle separately in the catalog.
This way, we get closer to the goal of supporting POCO objects as grains in the future, if all that is in `Grain` can be implemented by scoped activation services and lifetime observers.

File: src/Orleans/Core/Grain.cs (hunk 7 above)
Comment (jason-bragg) [reply src/Orleans/Core/Grain.cs]: hmm.. looks like there was  bug..?  We load the initial sate using what you describe, but the storage bridge was setup with a different name, meaning the initial loaded state is different from the state the bridge uses.  Not sure how that got past tests, but that's a serious bug, unless I'm misunderstanding what I'm reading.

File: src/OrleansRuntime/Catalog/Catalog.cs (hunk 8 above)
Comment (jason-bragg) [reply src/OrleansRuntime/Catalog/Catalog.cs]: The expected behavior is that state be loaded then pre-activate, activate, post-activate.  Since state load happens in lifecycle, this will need to be after onStart, for now.  When onActivateAsync is called by the lifecycle, this can be as well.

File: src/OrleansRuntime/Catalog/Catalog.cs (hunk 9 above)
Comment (jason-bragg) [reply src/OrleansRuntime/Catalog/Catalog.cs]: That is the plan.  All of grain lifecycle should be handled within the lifecycle framework.  We've just not refactored it all yet.  A good next step would be to add pre-activate, activate, and post-activate stages to the lifecycle, and move the onactivate/deavtivate within the lifecycle (as you've describe) along with the pre/post activate logic.  Though, I'm not convinced we need the pre/post activate, so I'd like to understand why that was needed.

File: src/Orleans/Core/Grain.cs (hunk 7 above)
Comment (jason-bragg) [reply src/Orleans/Core/Grain.cs]: Change was introduced in July of last year, so it may be lower impact than I'd expected.   https://github.com/dotnet/orleans/commit/f634bde01f8d8fac696c0086986a1bbb61815cd8
Anyone familiar with what this was for?  @ReubenBond? @jdom?

Comment (jason-bragg) [reply src/Orleans/Core/Grain.cs]: Ah.  I see.  This addressed a problem which only existed because the bridge was not being used to read the initial state.  Current code should be fine.

File: src/Orleans/LogConsistency/LogConsistentGrain.cs [hunk 10]
```diff
@@ -0,0 +1,96 @@
...
+        /// (in which case user MUST configure a consistency provider)
+        /// </summary>
+        protected abstract ILogViewAdaptorFactory DefaultAdaptorFactory { get; }
+
+        public virtual void Participate(ILifecycleObservable<GrainLifecycleStage> lifecycle)
+        {
+            lifecycle.Subscribe(GrainLifecycleStage.SetupState, OnSetupState);
+        }
+
+        Task OnSetupState(CancellationToken ct)
```
Comment (jdom) [inline src/Orleans/LogConsistency/LogConsistentGrain.cs]: nit: add the private modifier

File: src/OrleansRuntime/Catalog/Catalog.cs [hunk 11]
```diff
@@ -1376,13 +1178,16 @@ private void RerouteAllQueuedMessages(ActivationData activation, ActivationAddre
 
                 activationsFailedToActivate.Increment();
 
+                /// TODO: During lifecycle refactor discuss with team whether activation failure should have a well defined exception, or throw whatever
+                ///   exception caused activation to fail, with no indication that it occured durring activation
+                ///   rather than the grain call.
+                OrleansLifecycleCanceledException canceledException = exc as OrleansLifecycleCanceledException;
+                if(canceledException?.InnerException != null)
+                {
+                    throw canceledException.InnerException;
```
Comment (jdom) [inline src/OrleansRuntime/Catalog/Catalog.cs]: This would completely wipe out the original stack trace of the exception, which is very important for troubleshooting what went wrong. You might want to do the following instead:
```
ExceptionDispatchInfo.Capture(canceledException.InnerException).Throw();
```

File: test/TesterInternal/ErrorInjectionStorageProvider.cs [hunk 12]
```diff
@@ -53,6 +48,14 @@ public override string Message
                 return "ErrorInjectionPoint=" + Enum.GetName(typeof(ErrorInjectionPoint), errorInjectionPoint);
             }
         }
+
+#if !NETSTANDARD
```
Comment (jdom) [inline test/TesterInternal/ErrorInjectionStorageProvider.cs]: Please undo this change after rebasing on top of `master`, since we removed all NETSTANDARD conditional compilation statements in #3368 that was just merged

**Comment by jason-bragg** [comment]
@dotnet-bot test netfx-functional please

File: src/Orleans/Core/GrainExtensions.cs [hunk 13]
```diff
@@ -223,6 +228,20 @@ public static string GetPrimaryKeyString(this IGrainWithStringKey grain)
                 throw new InvalidOperationException("Invoking of methods with one way flag must result in completed task");
             }
         }
+        
+        public static IStorageProvider GetStorageProvider(this Grain grain, IServiceProvider services)
```
[the remark this answers is not in the document]
Comment (jdom) [reply src/Orleans/Core/GrainExtensions.cs]: At the very least I would constrain it to `Grain<T>`, not every `Grain`, but I think I prefer in something different than GrainExtensions that doesn't pollute the intellisense (or even not provide an extension method for this).
As you mentioned it's probably only going to be used by some infrastructure code (even just from us). And as we endorse facets, this is even less relevant.

**Comment by jason-bragg** [comment]
Squashed rebased.
Addressed all comments.

File: src/Orleans/Core/GrainExtensions.cs (hunk 13 above)
[the remark this answers is not in the document]
Comment (jdom) [reply src/Orleans/Core/GrainExtensions.cs]: probably missed addressing this one

[the remark this answers is not in the document]
Comment (jason-bragg) [reply src/Orleans/Core/GrainExtensions.cs]: Didn't fix because there was nothing broken.  Moved to namespace Orleans.Storage anyway, so it won't show up unless someone is already using storage features.

**Comment by jason-bragg** [comment]
ping

**Comment by jdom** [comment]
Oh, sorry, didn't merge as I thought you found a big issue with the name we use as state key. Was this resolved?

**Comment by ReubenBond** [comment]
Oh, I thought that was the existing behavior which was just being carried forwards

**Comment by jdom** [comment]
If that really existed, it would be a huge issue that needs to be investigated and potentially addressed asap (not just for 2.0). That's why I thought he was still investigating it, because it's crazy that no one encountered this before

**Comment by jdom** [comment]
@jason-bragg can you confirm?

**Comment by jason-bragg** [comment]
TL,DR: The original code was correct.  The new code is correct.

Full description of original problem: "Incomplete grainType when activating generic grains #1579"
Fix: "Fix bug with generic state parameter caused by inconsistent use of grainClassName / genericArgument / genericInterface #1897"

The difference in state name generation in the catalog vs. the bridge is due to the fact that the catalog used the type from the grain type data which did not have the generic value which is only known at runtime.  The bridge did not need to deal with this problem because it used the fully qualified runtime grain type.  By removing the (duplicate) load logic from the catalog, we no longer use the unqualified type from the grain type data, so the issue goes away.
